### PR TITLE
Add rendering support for Arm Control in the Emulator

### PR DESCRIPTION
**Pull request by TamRuas**
@EranGon @liadtal 

Opening separate tasks for: Order, Adding chart to display, and checking app insights. 
Thanks!

### CODE REVIEW

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 1]
```diff
@@ -6,12 +6,19 @@
...
 namespace Microsoft.Azure.Monitoring.SmartDetectors.MonitoringApplianceEmulator.ViewModels
 {
+    using System;
     using System.Collections.Generic;
     using System.Collections.ObjectModel;
+    using System.Diagnostics.CodeAnalysis;
     using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
     using Controls;
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >Controls [](start = 10, length = 8)

not your change, but can you put the full namespace here?

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 2]
```diff
@@ -49,10 +56,13 @@ public AlertDetailsControlViewModel()
         /// </summary>
         /// <param name="alert">The alert.</param>
         /// <param name="alertDetailsControlClosed">Handler for closing the details control.</param>
+        /// <param name="armClient">Support for ARM request.</param>
         [InjectionConstructor]
+
         public AlertDetailsControlViewModel(
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: redundant empty linr

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 3]
```diff
@@ -143,5 +148,115 @@ private set
         public CommandHandler CloseControlCommand { get; }
 
         #endregion
+
+        private async Task<List<DisplayableAlertProperty>> ComposeArmProperties(AzureResourceManagerRequestAlertProperty armProperty, IAzureResourceManagerClient armClient)
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >ComposeArmProperties [](start = 59, length = 20)

add comment please

Comment (EranGon) [reply src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: for all methods

---
In reply to: [316140382](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/72#discussion_r316140382) [](ancestors = 316140382)

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 4]
```diff
@@ -143,5 +148,115 @@ private set
...
+        {
+            var url = armProperty.AzureResourceManagerRequestUri;
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >url [](start = 16, length = 3)

consider removing this variable - might be clearer what's the meaning when you use it

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 5]
```diff
@@ -143,5 +148,115 @@ private set
...
+            IReadOnlyList<DisplayableAlertProperty> propertyRefs = armProperty.PropertiesToDisplay;
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >propertyRefs [](start = 52, length = 12)

also this

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 6]
```diff
@@ -143,5 +148,115 @@ private set
...
+            List<DisplayableAlertProperty> displayableArmProperties = new List<DisplayableAlertProperty>();
+            try
+            {
+                List<JObject> response = await armClient.ExecuteArmQueryAsync(url, CancellationToken.None);
+
+                foreach (IReferenceAlertProperty propertyRef in propertyRefs.OfType<IReferenceAlertProperty>())
+                {
+                    JToken propertyVal = response[0].SelectToken(propertyRef.ReferencePath);
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >JToken propertyVal = response[0].SelectToken(propertyRef.ReferencePath); [](start = 20, length = 72)

better check if the reference exists, and if not - add an "error property" just for this reference

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 7]
```diff
@@ -143,5 +148,115 @@ private set
...
+                            KeyValueAlertProperty displayKeyValue = new KeyValueAlertProperty(keyValueRef.PropertyName, keyValueRef.DisplayName, keyValueRef.Order, keyValueField);
+                            displayableArmProperties.Add(displayKeyValue);
+                            break;
+
+                        case TableReferenceAlertProperty tableRef:
+                            displayText = new TextAlertProperty(tableRef.PropertyName, tableRef.DisplayName, tableRef.Order, errorValue);
+                            displayableArmProperties.Add(displayText);
+                            break;
+                    }
+                }
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: maybe extract this to a method

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 8]
```diff
@@ -143,5 +148,115 @@ private set
...
+            return displayableArmProperties;
+        }
+
+        private async Task<ObservableCollection<DisplayableAlertProperty>> ComposeAlertProperties(IAzureResourceManagerClient armClient)
+        {
+            try
+            {
+                IEnumerable<Task<List<DisplayableAlertProperty>>> armPropertiesTasks = this.Alert.ContractsAlert.AlertProperties.OfType<AzureResourceManagerRequestAlertProperty>()
+                    .Select(armProperty => this.ComposeArmProperties(armProperty, armClient));
+                List<DisplayableAlertProperty> armProperties = (await Task.WhenAll(armPropertiesTasks)).SelectMany(props => props).ToList();
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: please add some comments inline to explain the flow

File: src/SDK/SmartDetectorsSDK/Arm/IAzureResourceManagerClient.cs [hunk 9]
```diff
@@ -87,5 +88,17 @@ public interface IAzureResourceManagerClient
         /// <param name="cancellationToken">The cancellation token.</param>
         /// <returns>>A <see cref="Task{TResult}"/>, running the current operation, returning a list of all items returned.</returns>
         Task<List<JObject>> ExecuteArmQueryAsync(ResourceIdentifier resource, string suffix, string queryString, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Executes an ARM GET request, using the relative path, and query string.
+        /// For example, the following call gets the list of databases for an SQL Server resource:
+        /// <code>
+        /// List&lt;JObject&gt; databases = await ExecuteArmQueryAsync(sqlResource, "/databases", "api-version=2017-10-01-preview", cancellationToken);
```
Comment (EranGon) [inline src/SDK/SmartDetectorsSDK/Arm/IAzureResourceManagerClient.cs]: >sqlResource, "/databases", "api-version=2017-10-01-preview [](start = 71, length = 58)

OMG!

File: src/SDK/SmartDetectorsShared/Clients/ExtendedAzureResourceManagerClient.cs [hunk 10]
```diff
@@ -304,9 +304,26 @@ public async Task<string> GetLogAnalyticsWorkspaceIdAsync(ResourceIdentifier res
...
+            // Create relative path
+            Uri relativePath = new Uri(resource.ToResourceId() + suffix + "?" + queryString, UriKind.Relative);
+            return await this.ExecuteArmQueryAsync(relativePath, cancellationToken);
+        }
+
+        /// <summary>
+        /// Executes an ARM GET request, using the relative path, and query string.
+        /// For example, the following call gets the list of databases for an SQL Server resource:
+        /// <code>
+        /// List&lt;JObject&gt; databases = await ExecuteArmQueryAsync(sqlResource, "/databases", "api-version=2017-10-01-preview", cancellationToken);
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Clients/ExtendedAzureResourceManagerClient.cs]: >await ExecuteArmQueryAsync(sqlResource, "/databases", "api-version=2017-10-01-preview [](start = 44, length = 85)

fix this comment as well

**Comment by EranGon** [comment]
>         /// List&lt;JObject&gt; databases = await ExecuteArmQueryAsync(sqlResource, "/databases", "api-version=2017-10-01-preview", cancellationToken);

fix this comment as well

---
Refers to: src/SDK/SmartDetectorsShared/Clients/ExtendedAzureResourceManagerClient.cs:298 in 0a2b081. [](commit_id = 0a2b08130453d898b2a1aee9d37c86ef170d61e0, deletion_comment = False)

File: src/SDK/SmartDetectorsShared/Clients/ExtendedAzureResourceManagerClient.cs (hunk 10 above)
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Clients/ExtendedAzureResourceManagerClient.cs]: >ExecuteArmQueryAsync(sqlResource, "/databases", "api-version=2017-10-01-preview", cancellationToken [](start = 50, length = 99)

fix this comment as well

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs (hunk 7 above)
Comment (TamRuas) [reply src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: I can extract the "non error" switch case into the same function if cleaner.
The code is the same except for the table. So i can add an error flag to the signature to differentiate.

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 11]
```diff
@@ -143,5 +146,138 @@ private set
...
+                                    .Select(value => value.ToDictionary(item => item.Key, item => item.Value))
+                                    .ToList();
+                                TableAlertProperty<Dictionary<string, JToken>> displayTable = new TableAlertProperty<Dictionary<string, JToken>>(tableRef.PropertyName, tableRef.DisplayName, tableRef.Order, true, tableRef.Columns, values);
+                                displayableArmProperties.Add(displayTable);
+                                break;
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >Exception [](start = 19, length = 9)

Please catch a specific exception type here, not any exception

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 12]
```diff
@@ -143,5 +147,138 @@ private set
...
+                List<DisplayableAlertProperty> displayableAlertProperties = this.Alert.ContractsAlert.AlertProperties.OfType<DisplayableAlertProperty>()
+                    .Where(prop => this.supportedPropertiesTypes.Contains(prop.Type))
+                    .Union(armProperties) // Combine all properties into a single list.
+                    .OrderBy(prop => prop.Order) // Reorder properies
+                    .ThenBy(prop => prop.PropertyName)
+                    .ToList();
+
+                return new ObservableCollection<DisplayableAlertProperty>(displayableAlertProperties);
+            }
+            catch (Exception e)
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >Exception [](start = 19, length = 9)

also here

File: src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs [hunk 13]
```diff
@@ -143,5 +147,138 @@ private set
         public CommandHandler CloseControlCommand { get; }
 
         #endregion
+
+        /// <summary>
+        /// Retrieves and composes Alert propeties, for an Arm request Alert property.
```
Comment (bermanguy) [inline src/SDK/MonitoringApplianceEmulator/ViewModels/AlertDetailsControlViewModel.cs]: >propeties [](start = 41, length = 9)

properties :) 
you can install Visual Studio Spell Checker

**Pull request merged**